### ISSUE TITLE
fit_generator consume too much ram memory if using large dataset

### ISSUE THREAD

**Issue by LCorleone** [keras-team/keras#11678]
hi all, this is my own data generator.
```
def img_generator(img_path_list, label, image_shape=(112, 112, 3), batchsize=32, if_shuffle=True):
    if if_shuffle:
        img_path_list, label = shuffle(img_path_list, label)

    print('total img is {}, and from total subject {}'.format(len(label), len(set(label))))
    # iter
    batch_x = np.zeros((batchsize,) + image_shape, dtype=K.floatx())
    batch_y = np.zeros((batchsize,), dtype=np.int32)
    count = 0
    while 1:
        del batch_x, batch_y
        batch_x = np.zeros((batchsize,) + image_shape, dtype=K.floatx())
        batch_y = np.zeros((batchsize,), dtype=np.int32)
        i_index = 0
        for index in range(count * batchsize, (count + 1) * batchsize):
            img = load_img(img_path_list[index],
                           grayscale=False,
                           target_size=(image_shape[0], image_shape[1]))
            x = img_to_array(img)
            batch_x[i_index] = x
            batch_y[i_index] = label[index]
            i_index += 1

        if count < math.floor(len(img_path_list) / float(batchsize)) - 1:
            count += 1
        else:
            if if_shuffle:
                img_path_list, label = shuffle(img_path_list, label)
            count = 0
        # print('here')
        # print('size of batch x: {}'.format(sys.getsizeof(batch_x)))
        yield (batch_x, batch_y)
```
And I train my model like this: Batch size is 4，very small.
```
path_list, label = gen_img_path_label(train_dir)
train_generator = img_generator(path_list, label, batchsize=batch_size, image_shape=(112, 112, 3))
num_train = len(label)
model.fit_generator(train_generator, epochs=num_epochs, steps_per_epoch=num_train // batch_size, callbacks=[checkpoint], max_queue_size=1)
```
it works fine when my image dataset has 10572 labels of 490000 images. it consumes 3.2G ram.
however, when I use a large dataset, has 73486 labels of 3740000 images. it consumes 28G ram totally!
and it gets stuck.
![image](https://user-images.githubusercontent.com/22718985/48710119-5b62d300-ec42-11e8-9667-623a6d435424.png)
my question is:
the code is the same, the only difference is the parameters in last dense layer gets much bigger because the label increases much. But, that doesn't matter. the generator only generates batch images, why it consumes so much ram for that I only change the dataset.
is there any ideas?

**Comment by Dref360** [comment]
The size of this list is quite huge (7x bigger than your other test).

Using your current code, the list only exists in the main process. (You use threads, not processes)

You could use an hdf5 file to store those paths (Warning, they are not thread-safe).
Or store those paths on your disk in a numpy format and read them using np.memmap.

**Comment by LCorleone** [comment]
@Dref360 
I tried to use the first 1/4 path_list and label, named path_list_temp and label_temp and delete the raw path_list and label to check the code. Nothing changed, it also consumes about 28G ram. So strange.

**Comment by LCorleone** [comment]
i use tf 1.10 
now，this warning shows up, i really have no idea for this. the ram usage becomes larger and larger and totally consumes all of my ram memory.
![image](https://user-images.githubusercontent.com/22718985/48764222-04183d80-ecea-11e8-8827-211fe554b5d9.png)

**Comment by Dref360** [comment]
You will have to profile it yourself and see where the memory is allocated.

**Comment by LCorleone** [comment]
@Dref360 thanks，finally i find something wrong in my model. close this issue.

**Comment by LSerafini** [comment]
Hi @LCorleone !
Could you please explain what was the cause of your issue?
I have a similar problem.

Thank you!
